apps/server/src/shared/domain/service/file-copy-append.service.ts

```diff
@@ -72,23 +76,151 @@ export class FileCopyAppendService {
+		return elements.find((el) => el.type === CopyElementType.FILE_GROUP) as CopyStatus;
 	}
 
-	private replaceFileGroup(elements: CopyStatus[] = [], fileGroupStatus: CopyStatus): CopyStatus[] {
-		return elements.map((el) => (el.type === CopyElementType.FILE_GROUP ? fileGroupStatus : el));
+	private setFileGroupStatus(elements: CopyStatus[] = [], updatedFileGroupStatus: CopyStatus): CopyStatus[] {
+		const fileGroupStatus = this.getFileGroupStatus(elements);
+		if (fileGroupStatus) {
+			return elements.map((el) => (el.type === CopyElementType.FILE_GROUP ? updatedFileGroupStatus : el));
+		}
+		return [...elements, updatedFileGroupStatus];
 	}
 
 	private createFileStatuses(files: FileDto[]): CopyStatus[] {
 		return files.map((file) => ({
 			type: CopyElementType.FILE,
-			name: file.name,
+			title: file.name,
 			status: CopyStatusEnum.SUCCESS,
 		}));
 	}
+
+	private createFileStatusesByCopyResult(files: FileLegacyResponse[]): CopyStatus[] {
+		return files.map(({ oldFileId, fileId, filename }) => ({
+			type: CopyElementType.FILE,
+			status: fileId ? CopyStatusEnum.SUCCESS : CopyStatusEnum.FAIL,
+			title: filename ?? `(old fileid: ${oldFileId})`,
+		}));
+	}
+
+	async copyFiles(copyStatus: CopyStatus, courseId: EntityId, userId: EntityId, jwt: string): Promise<CopyStatus> {
+		if (copyStatus.type === CopyElementType.LESSON) {
+			return this.copyEmbeddedFilesOfLessons(copyStatus, courseId, userId);
+		}
+		if (copyStatus.type === CopyElementType.TASK) {
+			const updatedStatus = await this.appendFilesToTask(copyStatus, jwt);
+			return this.copyEmbeddedFilesOfTasks(updatedStatus, courseId, userId);
+		}
+		if (copyStatus.elements && copyStatus.elements.length > 0) {
+			copyStatus.elements = await Promise.all(
+				copyStatus.elements.map((el) => this.copyFiles(el, courseId, userId, jwt))
+			);
+			copyStatus.status = this.copyHelperService.deriveStatusFromElements(copyStatus.elements);
+		}
+		return Promise.resolve(copyStatus);
+	}
+
+	async copyEmbeddedFilesOfLessons(
+		lessonCopyStatus: CopyStatus,
+		courseId: EntityId,
+		userId: EntityId
+	): Promise<CopyStatus> {
+		const lesson = lessonCopyStatus.copyEntity as Lesson;
+		const legacyFileIds: string[] = [];
+
+		lesson.contents.forEach((item: IComponentProperties) => {
+			if (item.content === undefined || !('text' in item.content)) {
+				return;
+			}
+
+			const contentFileIds = this.extractOldFileIds(item.content.text);
+
+			legacyFileIds.push(...contentFileIds);
+		});
+
+		if (legacyFileIds.length > 0) {
+			const fileCopyResults: FileLegacyResponse[] = await Promise.all(
+				legacyFileIds.map((fileId) => this.fileLegacyService.copyFile({ fileId, targetCourseId: courseId, userId }))
+			);
+
+			fileCopyResults.forEach(({ oldFileId, fileId, filename }) => {
+				lesson.contents = lesson.contents.map((item: IComponentProperties) => {
+					if ('text' in item.content && fileId && filename) {
```

Comment:
if(isTextLesson(lesson)) {

Reply:
Applying this change results in next line throwing a linter error.
I think it is a side effect of having optional properties in IComponentProperties. Maybe solvable through guardClause.